File: app/src/main/java/app/kamy/saatApp/domain/faraidh/FaraidhFraction.kt
package app.kamy.saatApp.domain.faraidh

import java.math.BigDecimal
import java.math.BigInteger
import java.math.RoundingMode

private fun BigInteger.lcm(other: BigInteger): BigInteger {
    if (this == BigInteger.ZERO || other == BigInteger.ZERO) return BigInteger.ZERO
    return this.multiply(other).abs().divide(this.gcd(other))
}

data class FaraidhFraction(
    val numerator: BigInteger,
    val denominator: BigInteger
) {
    init {
        require(denominator > BigInteger.ZERO) { "Denominator must be positive" }
    }

    fun normalized(): FaraidhFraction {
        val gcd = numerator.gcd(denominator)
        val den = if (denominator.signum() < 0) denominator.negate() else denominator
        val num = if (denominator.signum() < 0) numerator.negate() else numerator
        return FaraidhFraction(num / gcd, den / gcd)
    }

    fun add(other: FaraidhFraction): FaraidhFraction {
        val lcm = denominator.lcm(other.denominator)
        val a = numerator * (lcm / denominator)
        val b = other.numerator * (lcm / other.denominator)
        return FaraidhFraction(a + b, lcm).normalized()
    }

    fun multiplyScalar(count: Int): FaraidhFraction {
        if (count <= 0) return ZERO
        return FaraidhFraction(numerator * BigInteger.valueOf(count.toLong()), denominator).normalized()
    }

    fun divideAmongHeads(heads: Int): FaraidhFraction {
        if (heads <= 0) return ZERO
        return FaraidhFraction(numerator, denominator * BigInteger.valueOf(heads.toLong())).normalized()
    }

    fun toDecimal(scale: Int = 8): BigDecimal =
        BigDecimal(numerator).divide(BigDecimal(denominator), scale, RoundingMode.HALF_UP)

    fun toPercentage(scale: Int = 4): BigDecimal =
        toDecimal(scale + 2).multiply(BigDecimal(100)).setScale(scale, RoundingMode.HALF_UP)

    fun toDisplayString(): String {
        val n = normalized()
        return "${n.numerator}/${n.denominator}"
    }

    fun toCashAmount(estate: BigDecimal, scale: Int = 2): BigDecimal =
        estate.multiply(BigDecimal(numerator))
            .divide(BigDecimal(denominator), scale, RoundingMode.HALF_UP)

    companion object {
        val ZERO = FaraidhFraction(BigInteger.ZERO, BigInteger.ONE)
        val ONE = FaraidhFraction(BigInteger.ONE, BigInteger.ONE)

        fun of(n: Long, d: Long): FaraidhFraction = FaraidhFraction(
            BigInteger.valueOf(n),
            BigInteger.valueOf(d)
        ).normalized()

        fun sumOf(fractions: List<FaraidhFraction>): FaraidhFraction =
            fractions.fold(ZERO) { acc, f -> acc.add(f) }

        fun applyAwl(shares: List<Pair<HeirType, FaraidhFraction>>): List<Pair<HeirType, FaraidhFraction>> {
            val total = sumOf(shares.map { it.second })
            if (total.numerator <= total.denominator) return shares
            return shares.map { (type, frac) ->
                type to FaraidhFraction(
                    frac.numerator * total.denominator,
                    frac.denominator * total.numerator
                ).normalized()
            }
        }

        fun applyRadd(
            shares: List<Triple<HeirType, FaraidhFraction, Boolean>>,
            spouseTypes: Set<HeirType> = setOf(HeirType.HUSBAND, HeirType.WIFE)
        ): List<Pair<HeirType, FaraidhFraction>> {
            val total = sumOf(shares.map { it.second })
            if (total.numerator >= total.denominator) {
                return shares.map { it.first to it.second }
            }
            val surplus = FaraidhFraction(
                total.denominator - total.numerator,
                total.denominator
            )
            val eligible = shares.filter { (type, frac, _) ->
                type !in spouseTypes && frac.numerator > BigInteger.ZERO
            }
            if (eligible.isEmpty()) return shares.map { it.first to it.second }
            val eligibleTotal = sumOf(eligible.map { it.second })
            return shares.map { (type, frac, _) ->
                if (type in spouseTypes || frac.numerator == BigInteger.ZERO) {
                    type to frac
                } else {
                    val extraProp = FaraidhFraction(
                        frac.numerator * eligibleTotal.denominator,
                        frac.denominator * eligibleTotal.numerator
                    )
                    val extra = FaraidhFraction(
                        surplus.numerator * extraProp.numerator,
                        surplus.denominator * extraProp.denominator
                    ).normalized()
                    type to frac.add(extra).normalized()
                }
            }
        }
    }
}

File: iosApp/Saat/Domain/Models/FaraidhFraction.swift
import Foundation

struct FaraidhFraction: Comparable, Equatable, Hashable, Codable {
    let numerator: Int
    let denominator: Int
    
    init(numerator: Int, denominator: Int) {
        precondition(denominator > 0, "Denominator must be positive")
        self.numerator = numerator
        self.denominator = denominator
    }
    
    static func < (lhs: FaraidhFraction, rhs: FaraidhFraction) -> Bool {
        return lhs.numerator * rhs.denominator < rhs.numerator * lhs.denominator
    }
    
    static func == (lhs: FaraidhFraction, rhs: FaraidhFraction) -> Bool {
        return lhs.numerator * rhs.denominator == rhs.numerator * lhs.denominator
    }
    
    func hash(into hasher: inout Hasher) {
        let norm = normalized()
        hasher.combine(norm.numerator)
        hasher.combine(norm.denominator)
    }
    
    func normalized() -> FaraidhFraction {
        let divisor = gcd(numerator, denominator)
        let den = denominator < 0 ? -denominator : denominator
        let num = denominator < 0 ? -numerator : numerator
        return FaraidhFraction(numerator: num / divisor, denominator: den / divisor)
    }
    
    func add(_ other: FaraidhFraction) -> FaraidhFraction {
        let commonDenominator = lcm(denominator, other.denominator)
        let a = numerator * (commonDenominator / denominator)
        let b = other.numerator * (commonDenominator / other.denominator)
        return FaraidhFraction(numerator: a + b, denominator: commonDenominator).normalized()
    }
    
    func subtract(_ other: FaraidhFraction) -> FaraidhFraction {
        let commonDenominator = lcm(denominator, other.denominator)
        let a = numerator * (commonDenominator / denominator)
        let b = other.numerator * (commonDenominator / other.denominator)
        return FaraidhFraction(numerator: a - b, denominator: commonDenominator).normalized()
    }
    
    func multiply(_ other: FaraidhFraction) -> FaraidhFraction {
        return FaraidhFraction(numerator: numerator * other.numerator, denominator: denominator * other.denominator).normalized()
    }
    
    func divide(_ other: FaraidhFraction) -> FaraidhFraction {
        precondition(other.numerator != 0, "Division by zero fraction")
        return FaraidhFraction(numerator: numerator * other.denominator, denominator: denominator * other.numerator).normalized()
    }
    
    func multiplyScalar(_ count: Int) -> FaraidhFraction {
        if count <= 0 { return .zero }
        return FaraidhFraction(numerator: numerator * count, denominator: denominator).normalized()
    }
    
    func divideAmongHeads(_ heads: Int) -> FaraidhFraction {
        if heads <= 0 { return .zero }
        return FaraidhFraction(numerator: numerator, denominator: denominator * heads).normalized()
    }
    
    func toDecimal(scale: Int = 8) -> Decimal {
        let numDec = Decimal(numerator)
        let denDec = Decimal(denominator)
        return numDec / denDec
    }
    
    func toPercentage(scale: Int = 4) -> Decimal {
        let decimal = toDecimal(scale: scale + 2)
        return decimal * 100
    }
    
    func toDisplayString() -> String {
        let n = normalized()
        return "\(n.numerator)/\(n.denominator)"
    }
    
    func toCashAmount(estate: Decimal) -> Decimal {
        return estate * Decimal(numerator) / Decimal(denominator)
    }
    
    static let zero = FaraidhFraction(numerator: 0, denominator: 1)
    static let one = FaraidhFraction(numerator: 1, denominator: 1)
    
    static func of(numerator: Int, denominator: Int) -> FaraidhFraction {
        return FaraidhFraction(numerator: numerator, denominator: denominator).normalized()
    }
    
    static func sumOf(_ fractions: [FaraidhFraction]) -> FaraidhFraction {
        return fractions.reduce(.zero) { $0.add($1) }
    }
    
    static func applyAwl(shares: [(HeirType, FaraidhFraction)]) -> [(HeirType, FaraidhFraction)] {
        let total = sumOf(shares.map { $0.1 })
        if total.numerator <= total.denominator { return shares }
        return shares.map { (type, frac) in
            (type, FaraidhFraction(numerator: frac.numerator * total.denominator, denominator: frac.denominator * total.numerator).normalized())
        }
    }
    
    static func applyRadd(shares: [(HeirType, FaraidhFraction, Bool)], spouseTypes: Set<HeirType> = [.husband, .wife]) -> [(HeirType, FaraidhFraction)] {
        let total = sumOf(shares.map { $0.1 })
        if total.numerator >= total.denominator {
            return shares.map { ($0.0, $0.1) }
        }
        let surplus = FaraidhFraction(numerator: total.denominator - total.numerator, denominator: total.denominator)
        let eligible = shares.filter { !spouseTypes.contains($0.0) && $0.1.numerator > 0 }
        if eligible.isEmpty { return shares.map { ($0.0, $0.1) } }
        let eligibleTotal = sumOf(eligible.map { $0.1 })
        return shares.map { (type, frac, _) in
            if spouseTypes.contains(type) || frac.numerator == 0 {
                return (type, frac)
            } else {
                let extraProp = FaraidhFraction(numerator: frac.numerator * eligibleTotal.denominator, denominator: frac.denominator * eligibleTotal.numerator)
                let extra = FaraidhFraction(numerator: surplus.numerator * extraProp.numerator, denominator: surplus.denominator * extraProp.denominator).normalized()
                return (type, frac.add(extra).normalized())
            }
        }
    }
}

private func gcd(_ a: Int, _ b: Int) -> Int {
    var x = abs(a)
    var y = abs(b)
    while y != 0 {
        let temp = y
        y = x % y
        x = temp
    }
    return x == 0 ? 1 : x
}

private func lcm(_ a: Int, _ b: Int) -> Int {
    if a == 0 || b == 0 { return 0 }
    return abs(a * b) / gcd(a, b)
}
